File: 2501~3000/2742. Painting the Walls/main.py
"""
Source: https://rb.gy/ftxv2
Date: 2023/7/21
Skill: dp更新過程為何從現在推到未來？因為無法窮舉之前的可能
Ref:
Runtime: 7402 ms, faster than 5.07%
Memory Usage: 24.87 MB, less than 62.98%
Time complexity:
Space complexity:
Constraints:

"""

import math, sys
from typing import List, Optional
from collections import defaultdict, Counter, deque
from heapq import heapify, heappush, heappop, nsmallest
import heapq
import functools
from bisect import bisect_left, bisect_right


class Solution:
    def paintWalls(self, cost: List[int], time: List[int]) -> int:
        cost.insert(0, 0)
        time.insert(0, 0)
        sz, offset, res = len(cost), len(cost) + 1, sys.maxsize
        dp = [[sys.maxsize for _ in range(2 * offset)] for _ in range(sz)]
        dp[0][offset] = 0
        for i in range(0, sz - 1):
            for j in range(-sz, sz + 1):
                dp[i + 1][j - 1 + offset] = min(dp[i + 1][j - 1 + offset], dp[i][j + offset])
                jj = min(j + time[i + 1], sz)
                dp[i + 1][jj + offset] = min(dp[i + 1][jj + offset], dp[i][j + offset] + cost[i + 1])

        for i in range(sz + 1):
            res = min(res, dp[-1][offset + i])

        return res


if __name__ == "__main__":
    s = Solution()

File: 2501~3000/2742. Painting the Walls/main.cpp
/**
 * Source: t.ly/EwrcQ
 * Date: 2023/10/14
 * Skill:
 * Runtime: 290 ms, faster than 17.01% of C++ online submissions
 * Memory Usage: 91.50 MB, less than 65.68% of C++ online submissions
 * Time complexity: O(n)
 * Space complexity: O(n)
 * Constraints:
 *      1 <= cost.length <= 500
 */

#include <vector>
#include <iostream>
#include <algorithm>
#include <map>
#include <set>
#include <unordered_set>
#include <queue>
#include <numeric>
#include <tuple>
#include <stack>
#include <bitset>

using namespace std;

#define ll long long
typedef pair<int, int> pairs;
typedef unsigned long long ULL;
using PULL = pair<ULL, ULL>;


class Solution {
    int offset = 500;
    int dp[501][501 * 2];
public:
    int paintWalls(vector<int>& cost, vector<int>& time) {
        int sz = cost.size(), res = INT32_MAX;
        cost.insert(begin(cost), 0);
        time.insert(begin(time), 0);
        for (int i = 0; i <= sz; i++)
            for (int j = -sz; j <= sz; j++)
                dp[i][j + offset] = INT32_MAX / 2;
        dp[0][0 + offset] = 0;
        for (int i = 0; i < sz; i++) {
            for (int j = -sz; j <= sz; j++) {
                if (j + 1 <= sz)
                    dp[i+1][j + offset] = min(dp[i+1][j + offset], dp[i][j + 1 + offset]);
                dp[i+1][min(j + time[i + 1], sz) + offset] = min(dp[i+1][min(j + time[i + 1], sz) + offset], dp[i][j + offset] + cost[i + 1]);
            }
        }
        for (int i = 0; i <= sz; i++)
            res = min(res, dp[sz][i + offset]);
        return res;
    }
};


int main() {
    Solution s;
    vector<int> nums{3,3,2};
    vector<vector<int>> arr{{1,10}, {3,3}};
}